### ISSUE TITLE
about train loss

### ISSUE THREAD

**Issue by dickang** [CoinCheung/BiSeNet#224]
I train with custom dataset. The loss curve obtained from train is like this. The loss in the first half of train is basically unchanged. What is the reason?
![train_results_loss](https://user-images.githubusercontent.com/46447643/158591278-c337fcb2-4b0e-439d-ad1c-11e8ddf11b73.png)
configs:
`    model_type='bisenetv1',
    n_cats=2,
    num_aux_heads=2,
    lr_start=1e-2,
    weight_decay=5e-4,
    warmup_iters=1000,
    max_iter=80000,
    dataset='CityScapes',
    im_root='./datasets/cityscapes',
    train_im_anns='./datasets/cityscapes/train.txt',
    val_im_anns='./datasets/cityscapes/val.txt',
    scales=[0.75, 2.],
    cropsize=[1024, 1024],
    eval_crop=[1024, 1024],
    eval_scales=[0.5, 0.75, 1.0, 1.25, 1.5, 1.75],
    ims_per_gpu=8,
    eval_ims_per_gpu=2,
    use_fp16=True,
    use_sync_bn=False,
    respth='./res',`

evaluating the final model:
INFO evaluate.py(207): single mIOU is: 0.9175312519073486
INFO evaluate.py(219): single scale crop mIOU is: 0.9175312519073486
INFO evaluate.py(225): ms flip mIOU is: 0.9229568243026733
INFO evaluate.py(237): ms crop mIOU is: 0.9237711429595947
INFO train_amp.py(191): |   single_scale |   single_scale_crop |   ms_flip |   ms_flip_crop |
|----------------+---------------------+-----------+----------------|
|       0.917531 |            0.917531 |  0.922957 |       0.923771 |

**Comment by CoinCheung** [comment]
Hi,

If it is not an implementation problem, it is more likely to be relevant to your training settings. My intuition is that this is associated with learning rate. I would like to use grid search to determine training hyper-parameters. 

I rememer someone has summarized an "8 ass article" for Chinese company interview. I cannot find the link to that page but I am impressed by some opinions in it:
* when you see both training loss and validation loss are dropping, it means your learning rate is so small that the model cannot converge.
* when you see training loss is eliminating but validation loss is not, it means "overfitting". You model has fallen into some "local optimal point", which fits training set well but cannot generalize to validation set.
* when you see training loss is higher than validation loss, it means data in validation set is easier than training set. Maybe you have used too much strong regularization( e.g. mixup), or maybe your dataset is not split well. The validation set consists of easy samples which is not the case of training set.
* when you see training loss curve is smooth but validation loss curve fluctuate fiercely, it means validation dataset is not so representative. It can be caused by too few samples in the validation set. Or it can be caused by distribution difference between training and validation set.
* If you set learning rate too large, the model would skip over the "minimal loss point" back and forth, and the training loss would not descend.
* If you set learning rate too small, the model would converge slowly.

In your case, if we only consider traing loss curve, it can be caused by a large initial learning rate. Maybe you can use a small learning rate and try again.

**Comment by dickang** [comment]
> Hi,
> 
> If it is not an implementation problem, it is more likely to be relevant to your training settings. My intuition is that this is associated with learning rate. I would like to use grid search to determine training hyper-parameters.
> 
> I rememer someone has summarized an "8 ass article" for Chinese company interview. I cannot find the link to that page but I am impressed by some opinions in it:
> 
> * when you see both training loss and validation loss are dropping, it means your learning rate is so small that the model cannot converge.
> * when you see training loss is eliminating but validation loss is not, it means "overfitting". You model has fallen into some "local optimal point", which fits training set well but cannot generalize to validation set.
> * when you see training loss is higher than validation loss, it means data in validation set is easier than training set. Maybe you have used too much strong regularization( e.g. mixup), or maybe your dataset is not split well. The validation set consists of easy samples which is not the case of training set.
> * when you see training loss curve is smooth but validation loss curve fluctuate fiercely, it means validation dataset is not so representative. It can be caused by too few samples in the validation set. Or it can be caused by distribution difference between training and validation set.
> * If you set learning rate too large, the model would skip over the "minimal loss point" back and forth, and the training loss would not descend.
> * If you set learning rate too small, the model would converge slowly.
> 
> In your case, if we only consider traing loss curve, it can be caused by a large initial learning rate. Maybe you can use a small learning rate and try again.

Thank you. It's very helpful to me.

**Comment by CoinCheung** [comment]
Let me close this, since we have done with our discussions for now.
